lib/WUI/nhttp/req_parser.cpp

```diff
@@ -241,8 +270,159 @@ bool RequestParser::uri_filename(char *buffer, size_t buffer_size) const {
+    switch (status) {
+    case NonceStatus::Invalid:
+        return UnauthenticatedStatusPage(Status::Unauthorized, status_page_handling(), accepts_json, AuthMethod::Digest, false, fresh_nonce);
+    case NonceStatus::Stale: {
+        bool stale = check_digest_auth(recieved_nonce);
+        return UnauthenticatedStatusPage(Status::Unauthorized, status_page_handling(), accepts_json, AuthMethod::Digest, stale, fresh_nonce);
+    }
+    case NonceStatus::Valid: {
+        if (check_digest_auth(recieved_nonce)) {
+            return std::nullopt;
+        } else
+            return UnauthenticatedStatusPage(Status::Unauthorized, status_page_handling(), accepts_json, AuthMethod::Digest, false, fresh_nonce);
+    }
+    default:
+        return UnauthenticatedStatusPage(Status::Unauthorized, status_page_handling(), accepts_json, AuthMethod::Digest, false, fresh_nonce);
+    }
+}
+
+std::optional<UnauthenticatedStatusPage> RequestParser::authenticated_status() const {
+    // do we have api key?
+    if (!holds_alternative<monostate>(auth_status))
+        return authenticated_status_api_key();
+    else
+        return authenticated_status_digest();
+}
+
+namespace {
+    void md5_update(mbedtls_md5_context &ctx, string_view first) {
+        mbedtls_md5_update(&ctx, (const uint8_t *)first.data(), first.size());
+    }
+    template <typename... Args>
+    void md5_hash(uint8_t *hash, Args... args) {
+        mbedtls_md5_context ctx;
+        mbedtls_md5_init(&ctx);
+        mbedtls_md5_starts(&ctx);
+        (md5_update(ctx, args), ...);
+        mbedtls_md5_finish(&ctx, hash);
+        mbedtls_md5_free(&ctx);
+    }
+
+    bool method_to_string(Method method, char *buffer, size_t buffer_size) {
+        auto copy_name = [&](auto name) {
+            if (buffer_size <= strlen(name)) {
+                return false;
+            }
+
+            strlcpy(buffer, name, buffer_size);
+            return true;
+        };
+        switch (method) {
+        case Method::Get: {
+            const char *name = "GET";
+            return copy_name(name);
+        }
+        case Method::Delete: {
+            const char *name = "DELETE";
+            return copy_name(name);
+        }
+        case Method::Head: {
+            const char *name = "HEAD";
+            return copy_name(name);
+        }
+        case Method::Post: {
+            const char *name = "POST";
+            return copy_name(name);
+        }
+        case Method::Put: {
+            const char *name = "PUT";
+            return copy_name(name);
+        }
+        default:
+            return false;
+        }
+    }
+}
+
+bool RequestParser::check_digest_auth(uint64_t nonce_to_use) const {
+    if (response_size == 0)
+        return false;
+    char method_str[7];
+    if (!method_to_string(method, method_str, sizeof(method_str)))
+        return false;
+
+    uint8_t hash_a1[16];
+    md5_hash(hash_a1, "user:", "Printer API:", server->get_api_key());
```

Comment:
- It is also used in `UnauthenticatedStatusPage` in returning the Unauthorized header. I agree it would be nicer in a constant, but I'm not sure how to put it in this `const char*` in `status_page.cpp:83` without breaking the const, any idea?
```
const char *api_key_header = "WWW-Authenticate: ApiKey realm=\"Printer API\"\r\n";
```
- Checking more than one user name probably would not be a problem from the code perspective, we would just need to compute the hash more than once. But accepting more usernames and especially an empty one sounds really hacky to me from the user perspective, is it really something we want?